llama : KV cache view API + better KV cache management

ggerganov opened the pull request:
This should help to resolve some cases where the KV cache is not utilized optimally.

We also expose an API for observing the contents of the KV cache, which can be useful for debugging purposes.

ggerganov committed:
llama : keep track of used KV cells + better KV cache management

WeirdConstructor commented:
`llama_kv_cache_clear` doesn't decrease `used`, is that correct?

ggerganov committed:
llama : zero KV cache used upon clear

ggml-ci

KerfuffleV2 commented:
I skimmed through the code and it looks reasonable, but unfortunately I won't have a chance to actually test/look at it more in-depth until tomorrow morning. (~14h from now). Not sure if you want to wait.

KerfuffleV2 commented:
@ggerganov 

I made #4180 - may help you visualize visualize how these changes affect the structure of the KV cache cells. So far I haven't found a problem when testing this pull. The change I linked also compares the in-KV cache cell used count to what it calculates from walking the cells and it seems like they match.

I didn't actually compare how it looks with these changes vs without. Also still need to do a bit more double checking that stuff like doing a shift works also and the count remains correct.

KerfuffleV2 reviewed (approved):
As far as I can see this doesn't break stuff and tracking the used slots is good. As far as the changes in logic, I'd say it's inconclusive whether they actually help.

I ran a number of tests with `parallel -np 4 -cb -ns 50 -c 780`

For context (this part stayed the same between runs):

```plaintext
main: n_parallel = 4, n_sequences = 50, cont_batching = 1, system tokens = 293
Total prompt tokens:    782, speed: 12.57 t/s
Total gen tokens:      4129, speed: 66.35 t/s
Total speed (AVG):           speed: 78.92 t/s
```

Then I tested running `parallel` with 5 different seeds with `master` and with this pull. This is the result ("fail" in the misses column if the run didn't complete due to failing to find a slot):

|seed|master misses|pull misses|
|-|-|-|
|123|1445|1590|
|124|fail|2964|
|125|1623|2000|
|126|1917|2324|
|127|1230|1554|

A sample size of 5 isn't necessarily conclusive. The pull passed one run where `master` didn't, but generally it seemed like `master` avoided more cache misses. _If_ we can draw any conclusion from these tests, we could possibly say `master`'s behavior works better in the case where cache pressure isn't extreme, but the pull handles the extreme case better? But I'm not sure there's enough information to really have much confidence in making inferences like that.

KerfuffleV2 commented:
I tried
```c++
static bool llama_kv_cache_find_slot(
           struct llama_kv_cache & cache,
        const struct llama_batch & batch) {
    const uint32_t n_ctx    = cache.size;
    const uint32_t n_tokens = batch.n_tokens;

    if (n_tokens > n_ctx) {
        LLAMA_LOG_ERROR("%s: n_tokens=%d > n_ctx=%d\n", __func__, n_tokens, n_ctx);
        return false;
    }

    std::vector<llama_kv_cell> & kv_cells = cache.cells;
    int32_t curr_contig_idx = -1;
    uint32_t candidate_contig = n_ctx + 1;
    int32_t candidate_contig_idx = -1;

    for (int32_t i = 0; i < int32_t(cache.size); i++) {
        if (kv_cells[i].seq_id.empty()) {
            if (curr_contig_idx < 0) {
                curr_contig_idx = i;
            }
            continue;
        }
        if (curr_contig_idx < 0) {
            continue;
        }
        const uint32_t contig_size = uint32_t(i - curr_contig_idx);
        if (contig_size >= n_tokens && contig_size < candidate_contig) {
            candidate_contig_idx = curr_contig_idx;
            candidate_contig = contig_size;
        }
        curr_contig_idx = -1;
    }
    if (curr_contig_idx >= 0) {
        const uint32_t contig_size = uint32_t(kv_cells.size() - curr_contig_idx);
        if (contig_size >= n_tokens && contig_size < candidate_contig) {
            candidate_contig_idx = curr_contig_idx;
            candidate_contig = contig_size;
        }
    }
    if (candidate_contig_idx < 0 || candidate_contig < n_tokens) {
        //LLAMA_LOG_ERROR("%s: failed to find a slot for %d tokens\n", __func__, n_tokens);
        cache.head = 0;
        return false;
    }

    // printf("Found: %d @ %d (needed %d)\n", candidate_contig, candidate_contig_idx, n_tokens);
    cache.head = candidate_contig_idx;
    for (uint32_t i = 0; i < n_tokens; i++) {
        kv_cells[cache.head + i].pos = batch.pos[i];

        for (int32_t j = 0; j < batch.n_seq_id[i]; j++) {
            kv_cells[cache.head + i].seq_id.insert(batch.seq_id[i][j]);
        }
    }

    cache.used += n_tokens;
    return true;
}
```

but it actually seemed worse than this pull or the original behavior.

I guess the reason is because even if you use a smaller slot later on, a sequence can end leaving the remaining stuff in that smaller slot as an island. So I guess preferring the lowest possible index that has enough space is probably preferable, at least for the the pattern `parallel` uses.

KerfuffleV2 committed:
llama : allow exporting a view of the KV cache (#4180)

* Allow exporting a view of the KV cache

* Allow dumping the sequences per cell in common

* Track max contiguous cells value and position as well

* Fix max contiguous empty cells index calculation

Make dump functions deal with lengths or sequences counts > 10 better

* Fix off by one error in dump_kv_cache_view

* Add doc comments for KV cache view functions

Eliminate cell sequence struct; use llama_seq_id directly

Minor cleanups

ggerganov commented:
Yeah, the logic is not perfect, but I also don't want to make it too complex either. Hopefully, we will figure out how to defragment the cache in the future.

KerfuffleV2 commented on examples/parallel/parallel.cpp:
```diff
@@ -201,6 +203,9 @@ int main(int argc, char ** argv) {
     LOG_TEE("Processing requests ...\n\n");
 
     while (true) {
+        llama_kv_cache_view_update(ctx, &kvc_view);
+        dump_kv_cache_view_seqs(kvc_view, 40);
+
```
This stuff needs to go, or at least be controlled by a commandline flag. (I think it would be reasonable to just revert this to the `master` version since trying to add new options is such a massive pain.)

ggerganov replied:
> since trying to add new options is such a massive pain.

How come? It's quite easy I think

KerfuffleV2 replied:
Haha, okay, you got me there. I guess I meant to add them in a way that it's consistent from a user perspective.

The problem is I can do `./main --help` and see the `--dump-kv-cache` option and there's absolutely no way to know it only works for `parallel`. For that reason, I've been very reluctant to try to add any options.

KerfuffleV2 replied:
BTW, tested the updated version (lightly), seemed okay, so my initial approval still applies.

ggerganov replied:
Yeah, that's a small issue, but nothing serious. Btw, we can add this options to `main` - could be useful with loading state from file.

ggerganov committed:
common : add -dkvc arg for enabling kv cache dumps
